src/linker/rule.py

```diff
@@ -62,23 +64,40 @@ class ImplementedRule(Rule):
     execution_input: List[str]
     validation: str
     output: List[str]
+    resources: Optional[dict]
     envvars: dict
     diagnostics_dir: str
     image_path: str
     script_cmd: str
 
     def _build_rule(self) -> str:
-        return (
-            f"""
+        return self._build_io() + self._build_resources() + self._build_shell_command()
+
+    def _build_io(self) -> str:
+        return f"""
 rule:
     name: "{self.name}"
+    message: "Running Implementation {self.name}"
```

Comment:
This is a nit, but I think for our users, this should include the name of the _step_ as well (and not capitalize "implementation").

Something like "Running {step_name} implementation: {self.name}"

Reply:
done